### PR TITLE
add celery tasks, now call these from models

### PR DESCRIPTION
**Pull request by StephenCoady**
fyi @nphilipp 

Signed-off-by: Stephen Coady <scoady@redhat.com>

### CODE REVIEW

File: bodhi/server/models.py [hunk 1]
```diff
@@ -2390,18 +2389,7 @@ def edit(cls, request, data):
...
-                    side_tag_pending_signing = up.release.get_pending_signing_side_tag(up.from_tag)
-                    koji.tagBuild(side_tag_pending_signing, build)
-                elif up.release.pending_signing_tag:
-                    # Add the release's pending_signing_tag to all new builds
-                    koji.tagBuild(up.release.pending_signing_tag, build)
-                else:
-                    # EL6 doesn't have these, and that's okay...
-                    # We still warn in case the config gets messed up.
-                    log.warning('%s has no pending_signing_tag' % up.release.name)
+            tag_build(new_builds, up)
```
Comment (abompard) [inline bodhi/server/models.py]: Celery tasks are called using the `.delay()` method, please check out how it's done with the `handle_update()` task.

File: bodhi/server/services/updates.py [hunk 2]
```diff
@@ -567,22 +568,7 @@ def new_update(request):
...
-                    if not koji.getTag(side_tag_signing_pending):
-                        koji.createTag(side_tag_signing_pending, parent=from_tag)
-                    if not koji.getTag(side_tag_testing_pending):
-                        koji.createTag(side_tag_testing_pending, parent=from_tag)
-                        koji.editTag2(side_tag_testing_pending, perm="autosign")
-
-                    to_tag = side_tag_signing_pending
-                    # Move every new build to <from_tag>-signing-pending tag
-                    update.add_tag(to_tag)
+                create_tag(updates, from_tag)
```
Comment (abompard) [inline bodhi/server/services/updates.py]: Same comment here, you'll have to use the `.delay()` method.

File: bodhi/server/tasks/__init__.py [hunk 3]
```diff
@@ -83,3 +83,50 @@ def handle_update(api_version: int, **kwargs):
     _do_init()
     handler = UpdatesHandler()
     handler.run(api_version=api_version, data=kwargs)
+
+
+@app.task(name="create_tag")
+def create_tag(updates: list, from_tag):
```
Comment (abompard) [inline bodhi/server/tasks/__init__.py]: Is `from_tag` supposed to be a string? If so, please type it accordingly.

File: bodhi/server/tasks/__init__.py [hunk 4]
```diff
@@ -83,3 +83,50 @@ def handle_update(api_version: int, **kwargs):
...
+            koji.createTag(side_tag_testing_pending, parent=from_tag)
+            koji.editTag2(side_tag_testing_pending, perm="autosign")
+
+        to_tag = side_tag_signing_pending
+        # Move every new build to <from_tag>-signing-pending tag
+        update.add_tag(to_tag)
+
+
+@app.task(name="tag_build")
+def tag_build(builds: list, update):
```
Comment (abompard) [inline bodhi/server/tasks/__init__.py]: Please type the `update` argument.

**Review by abompard — DISMISSED** [review]
Once the functions are moved to tasks, I'll expect some unit tests to fail (because the corresponding work would be done accordingly). I think you'll also have to reorder those parts of the unit tests, possibly create tests for the new tasks in the proper test module.

One more thing, I think it's worth adding an entry in the release notes saying that these actions will now be done asynchronously, please add a news fragment.

**Comment by mattiaverga** [comment]
@abompard should these two new tasks be worked by the `has_koji_mount` celery queue?
https://github.com/fedora-infra/bodhi/blob/983f7c74e832a37cc1c842937208d5536bad9864/celeryconfig.py#L19-L24

**Comment by abompard** [comment]
> @abompard should these two new tasks be worked by the `has_koji_mount` celery queue?

Unless I'm mistaken this code is only using the Koji API, so it can be executed from anywhere, we don't need the actual mount.

**Comment by nphilipp** [comment]
The type hints introduced a cyclic import (models -> tasks -> models) for which I have a fix locally. I noticed, however, that (expectedly) some existing tests block on when calling the new tasks. I'll mock these calls out and instead add unit tests testing `create_tag`, `task_build` in isolation modeled after existing tests for Celery tasks.

**Comment by nphilipp** [comment]
This fixes the tests breaking because of the introduction of using Celery tasks for Koji tagging. Needs unit tests for the new Celery tasks, though, so still WIP.

File: bodhi/tests/server/tasks/test_handle_side_and_related_tags.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+from unittest.mock import patch
+
+from bodhi.server import buildsys, models
+from bodhi.server.tasks import handle_side_and_related_tags_task
+from bodhi.server.tasks.handle_side_and_related_tags import main as hdle_side_and_related_tags_main
```
Comment (nphilipp) [inline bodhi/tests/server/tasks/test_handle_side_and_related_tags.py]: Whoa that's a long line. Perhaps shorten the name to `handle_srtags_main` here and below (as you're shortening `handle` anyway to make it fit)?

File: bodhi/tests/server/tasks/test_handle_side_and_related_tags.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+class TestTask(BasePyTestCase):
+    """Test the task in bodhi.server.tasks."""
+
+    @patch("bodhi.server.tasks.buildsys")
+    @patch("bodhi.server.tasks.initialize_db")
+    @patch("bodhi.server.tasks.config")
+    @patch("bodhi.server.tasks.handle_side_and_related_tags.main")
+    def test_task(self, main_function, config_mock, init_db_mock, buildsys):
+
```
Comment (nphilipp) [inline bodhi/tests/server/tasks/test_handle_side_and_related_tags.py]: Hmm, that blank line looks weird. Did you have some code previously and remove it later?

File: bodhi/tests/server/tasks/test_tag_update_builds.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+class TestTask(BasePyTestCase):
+    """Test the task in bodhi.server.tasks."""
+
+    @patch("bodhi.server.tasks.buildsys")
+    @patch("bodhi.server.tasks.initialize_db")
+    @patch("bodhi.server.tasks.config")
+    @patch("bodhi.server.tasks.tag_update_builds.main")
+    def test_task(self, main_function, config_mock, init_db_mock, buildsys):
+
```
Comment (nphilipp) [inline bodhi/tests/server/tasks/test_tag_update_builds.py]: Same blank line here. I don't think there are explicit rules but it looks weird and is inconsistent with the other methods.

File: bodhi/server/tasks/__init__.py [hunk 8]
```diff
@@ -28,7 +28,7 @@
 from bodhi.server.util import pyfile_to_module
 
 if typing.TYPE_CHECKING:
-    from bodhi.server.models import Update
+    from bodhi.server.models import Update  # pragma: no cover
```
Comment (nphilipp) [inline bodhi/server/tasks/__init__.py]: This could go to `if typing.TYPE_CHECKING:` above, so any other "typing-only" imports are covered.

**Review by nphilipp — APPROVED** [review]
Looks good to me!

**Pull request merged**